### PR TITLE
LIP0042: Change the type of IDs and add type hinting

### PR DESCRIPTION
**Pull request by sergeyshemyakov**

### CODE REVIEW

File: proposals/lip-0042.md [hunk 1]
```diff
@@ -86,18 +86,10 @@ The Reward module has the following internal function.
...
-
-##### Returns
-
-The reward for the block as a 64-bit unsigned integer.
-
 ##### Execution
 
 ```python
-getBlockReward(blockHeader):
+getBlockReward(blockHeader: BlockHeader) -> uint64:
```
Comment (ricott1) [inline proposals/lip-0042.md:92]: I think it would still be nice to define the `BlockHeader` type, similar to what we did here: https://github.com/LiskHQ/lips/pull/143/files#diff-855d735f598bef5639924d04e64a28a025082eec20f835526d33f68c5273625fR59

File: proposals/lip-0042.md [hunk 2]
```diff
@@ -108,7 +100,7 @@ getBlockReward(blockHeader):
   return defaultReward
 ```
 
-Here, `/` represents integer division, `isValidSeedReveal` is the function exposed by the [random module][research:random-module] and `impliesMaximalPrevotes` is the function exposed by the [BFT module][research:bft-module#specs].
+Here `isValidSeedReveal` is the function exposed by the [random module][research:random-module] and `impliesMaximalPrevotes` is the function exposed by the [BFT module][research:bft-module#specs].
```
Comment (ricott1) [inline proposals/lip-0042.md:103]: Why did you remove the reamrk about the integer division? In python 3 `/` represents true (float) division. If you want to do floor division (or integer division) we should use `//`: https://stackoverflow.com/questions/19507808/python-3-integer-division

Comment (ricott1) [inline proposals/lip-0042.md:103]: In other modules we used the notation `random.isValidSeedReveal` and `bft.impliesMaximalPrevotes`. It is also fine like this for me, but you may wanna consider this alternative.

**Review by ricott1 — CHANGES_REQUESTED** [review]

**Review by janhack — CHANGES_REQUESTED** [review]
The changes look good. I have a few further comments:

- It would be nice to use capitalization for modules names everywhere, in particular, for the Token module and Random module.
- For the minted reward, are we assuming the Token module emits an event or would it be better to emit an event from the Reward module here? I think it could be nice to emit an event in any case, also explaining why the rerward is 0 or was reduced by a factor of 4. Not sure how to best do it, whether to introduce multiple events or just one event with an integer property to distinguish between the cases.

File: proposals/lip-0042.md (hunk 2 above)
Comment (sergeyshemyakov) [reply proposals/lip-0042.md:103]: Forgot this detail, changed back.

Comment (sergeyshemyakov) [reply proposals/lip-0042.md:103]: I will just leave it like this.

File: proposals/lip-0042.md [hunk 3]
```diff
@@ -27,17 +27,27 @@ The goal of this LIP is to modularize the rewards assignment logic when a new bl
 
 The Reward module induces specific logic after the application of a block:
 
-* After applying the block, a certain quantity of the native token of the blockchain is minted and assigned to the block generator. The exact amount assigned to the block generator, i.e., the reward, is calculated depending on the rules of the [random module][research:random-module] and the [BFT module][research:bft-module].
+* After applying the block, a certain quantity of the native token of the blockchain is minted and assigned to the block generator. The exact amount assigned to the block generator, i.e., the reward, is calculated depending on the rules of the [random module][research:random-module] and the [BFT module][bft-module].
```
[the remark this answers is not in the document]
Comment (sergeyshemyakov) [reply proposals/lip-0042.md]: Fixed.

File: proposals/lip-0042.md [hunk 4]
```diff
@@ -27,17 +27,27 @@ The goal of this LIP is to modularize the rewards assignment logic when a new bl
...
+|**Name**   | **Type**  |   **Validation**  |   **Description** |
+|-----------|-----------|-------------------|-------------------|
+|BlockHeader   | object  |   see [LIP 0055][block-schema-lip]  |   The header of a block. |
+
 ### Constants
 
 | **Name**                      | **Type** | **Value**                                 |
 |-------------------------------|----------|-------------------------------------------|
 | `MODULE_ID_REWARD `           | bytes    |       TBD                                 |
+| `EVENT_TYPE_REWARD`           | bytes    |       0x0001                              |
```
[the remark this answers is not in the document]
Comment (sergeyshemyakov) [reply proposals/lip-0042.md]: Changed.

File: proposals/lip-0042.md [hunk 5]
```diff
@@ -86,21 +128,26 @@ The Reward module has the following internal function.
...
+* `amount`: amount of block reward to be minted.
+* `reduction`: an integer indicating whether the reward is reduced. Possible values are `REWARD_NO_REDUCTION`, `REWARD_REDUCTION_SEED_REVEAL`, `REWARD_REDUCTION_MAX_PREVOTES`.
+
 ##### Execution
 
 ```python
-getBlockReward(blockHeader: BlockHeader) -> uint64:
-  if isValidSeedReveal(blockHeader.generatorAddress, blockHeader.seedReveal) == false:
-    return 0
+getBlockReward(blockHeader: BlockHeader) -> tuple[uint64, uint32]:
```
[the remark this answers is not in the document]
Comment (sergeyshemyakov) [reply proposals/lip-0042.md]: Added.

File: proposals/lip-0042.md [hunk 6]
```diff
@@ -126,11 +173,17 @@ The default reward of the block as a 64-bit unsigned integer.
 
 #### After Block Execution
 
-After a block `b` is executed, the `mint` function exposed by the [token module][research:token-module] to assign the reward to the block generator as:
+After a block `b` is executed, the `mint` function exposed by the [Token module][token-module] to assign the reward to the block generator as:
 
 ```python
-blockReward = getBlockReward(b.header)
+(blockReward, reduction) = getBlockReward(b.header)
```
[the remark this answers is not in the document]
Comment (sergeyshemyakov) [reply proposals/lip-0042.md:179]: Yeah, we need the parenthesis because this is a tuple unpacking.

File: proposals/lip-0042.md [hunk 7]
```diff
@@ -126,11 +173,17 @@ The default reward of the block as a 64-bit unsigned integer.
...
 mint(b.header.generatorAddress, TOKEN_ID_REWARD, blockReward)
+
+eventData = {
```
[the remark this answers is not in the document]
Comment (sergeyshemyakov) [reply proposals/lip-0042.md]: Now call the `encode` function.

File: proposals/lip-0042.md [hunk 8]
```diff
@@ -78,6 +88,38 @@ This module does not define any state store.
 
 This module does not define any command.
 
+### Events
+
+#### RewardMinted
+
+Event is emitted when the reward is minted, it carries information about reward reduction. This event has `typeID = TYPE_ID_REWARD_MINTED`.
```
Comment (janhack) [inline proposals/lip-0042.md:95]:
```suggestion
The event is emitted when the reward is minted. In case of a reward reduction, it provides information about the reason for the reduction. This event has `typeID = TYPE_ID_REWARD_MINTED`.
```

File: proposals/lip-0042.md [hunk 9]
```diff
@@ -134,11 +173,19 @@ The default reward of the block as a 64-bit unsigned integer.
 
 #### After Block Execution
 
-After a block `b` is executed, the `mint` function exposed by the [token module][research:token-module] to assign the reward to the block generator as:
+After a block `b` is executed, the `mint` function exposed by the [Token module][token-module] to assign the reward to the block generator as:
 
 ```python
-blockReward = getBlockReward(b.header)
+(blockReward, reduction) = getBlockReward(b.header)
 mint(b.header.generatorAddress, TOKEN_ID_REWARD, blockReward)
```
Comment (janhack) [inline proposals/lip-0042.md:180]: Should we maybe only call mint here if the reward is positive?

File: proposals/lip-0042.md [hunk 10]
```diff
@@ -134,11 +173,19 @@ The default reward of the block as a 64-bit unsigned integer.
...
+
+eventDataBytes = encode(
+schema = rewardMintedDataSchema,
```
Comment (janhack) [inline proposals/lip-0042.md:183]: For readability, I would recommend some indentation here. You can even use `autopep8` for having automatic Python formatting.

**Review by janhack — CHANGES_REQUESTED** [review]
Overall looks good. I just have a few small points.

File: proposals/lip-0042.md [hunk 11]
```diff
@@ -134,11 +173,19 @@ The default reward of the block as a 64-bit unsigned integer.
...
+object = {
+    "amount": blockReward,
+    "reduction": reduction
+})
+emitEvent(MODULE_ID_REWARD, TYPE_ID_REWARD_MINTED, eventDataBytes, [b.header.generatorAddress])
```
Comment (ricott1) [inline proposals/lip-0042.md:188]: check the new API, we don't need to encode before anymore (see PR of 44 as reference)

**Review by ricott1 — CHANGES_REQUESTED** [review]

File: proposals/lip-0042.md (hunk 9 above)
Comment (sergeyshemyakov) [reply proposals/lip-0042.md]: Good point, I added a check.

File: proposals/lip-0042.md (hunk 10 above)
Comment (sergeyshemyakov) [reply proposals/lip-0042.md]: I think, my editor is not recognizing the code snippet in markdown as python, but I did the indentation manually.

File: proposals/lip-0042.md (hunk 11 above)
Comment (sergeyshemyakov) [reply proposals/lip-0042.md]: Super cool API!

File: proposals/lip-0042.md (hunk 8 above)
Comment (sergeyshemyakov) [reply proposals/lip-0042.md]: Changed as per suggestion.

**Review by ricott1 — APPROVED** [review]

**Review by janhack — APPROVED** [review]
Looks good, nice work!

**Review by karmacoma — APPROVED** [review]
